### PR TITLE
feat: Hyperjump arrival distance variable

### PR DESCRIPTION
**Pull request by petervdmeer**
Allow entering systems at additional distance from the target.

This implements the (arrival only) feature as requested in #4334 and detailed in the RFC [here](https://github.com/EndlessSkyCommunity/endless-sky/pull/51).

### CODE REVIEW

**Comment by Amazinite** [comment]
And how about a better keyword like `"arrival distance"`? Then the other half of #4334 could be the `"departure distance"`. (And actually looking at #4334 again, those are exactly the terms that the issue uses for this mechanic.) If I saw `link-shortfall` I'd have no idea what it meant.

**Comment by Tadrix** [comment]
How about just `arrival` and `departure`? Aren't two word keywords worse than one?

**Comment by Amazinite** [comment]
That works too, as we don't have `"habitable zone"` or `"asteroid belt"`, we just have `habitable` and `belt`.

**Comment by petervdmeer** [comment]
I changed they keyword to `arrival` as suggested.

Having a global keyword as suggested by Pointedstick would indeed be a nice service for plugin authors and for projects like EV conversion. I like to keep that for a separate PR.

The reason for me to implement this was this remark in #4334 :

> I was looking at ways of making combat more engaging for the player, and one of the issues is all the ships just thrown on top of each other.

With this change it is already possible for any player to experiment if the arrival distance will make combat more attractive.

For people that want to use this: @accursedCursive made a nice remark in #3628 that 1000 seems to be the sweet spot between more engaging combat and not being a nuisance in systems that have little combat and I think he has a good point there.

I also tried 10000 when I tested this feature, but that really gives one the feeling of traveling all alone in empty space.

**Comment by petervdmeer** [comment]
@Pointedstick Is there also a way to change "Changes requested" to something like "Changes done, awaiting re-review?"

**Comment by Tadrix** [comment]
Try 1250. That distance is notable as a "nominal point" for fuel collection values.

**Comment by jafdy** [comment]
Later, this isn't a bug fix.

**Comment by Amazinite** [comment]
Slapping that milestone on it and calling it a day.
If it works though, it's a +1 from me. All for giving plugin makers more tools.

File: source/System.cpp [hunk 1]
```diff
@@ -268,6 +268,8 @@ void System::Load(const DataNode &node, Set<Planet> &planets)
 			trade[value].SetBase(child.Value(valueIndex + 1));
 		else if(key == "object")
 			LoadObject(child, planets);
+		else if(key == "arrival" && child.Size() >= 2)
+			arrivalDistance = child.Value(1);
```
Comment (Amazinite) [inline source/System.cpp]: It looks like this should be `arrivalDistance = child.Value(valueIndex);` so that events can properly add to and remove from the arrival distance value.

**Comment by Zitchas** [comment]
Once this gets merged, my first order of business will be making a plugin that alters every system in the game so that all ships warp in a suitable distance out. Especially all those systems that have regular battles (and even more so, every system that has big mission battles).

Thanks to everyone working on this!

File: source/System.cpp [hunk 2]
```diff
@@ -268,6 +268,8 @@ void System::Load(const DataNode &node, Set<Planet> &planets)
 			trade[value].SetBase(child.Value(valueIndex + 1));
 		else if(key == "object")
 			LoadObject(child, planets);
+		else if(key == "arrival" && child.Size() >= 2)
```
Comment (tehhowch) [inline source/System.cpp]: we're guaranteed that the key has a value by line 184
(note that the "trade" key above accesses additional indices, which is why it needs the extra `child.Size()` check)

**Comment by Zitchas** [comment]
Actually, I like the idea that this change (if implemented widely) would actually make the JD more different from the current HD. Gives the two more tactical differences.

File: source/System.cpp [hunk 3]
```diff
@@ -268,6 +268,8 @@ void System::Load(const DataNode &node, Set<Planet> &planets)
 			trade[value].SetBase(child.Value(valueIndex + 1));
 		else if(key == "object")
 			LoadObject(child, planets);
+		else if(key == "arrival")
+			arrivalDistance = child.Value(valueIndex);
```
Comment (tehhowch) [inline source/System.cpp]: Should we allow negative values of `arrivalDistance`?

File: source/Ship.cpp [hunk 4]
```diff
@@ -1215,6 +1215,7 @@ void Ship::Move(vector<Visual> &visuals, list<shared_ptr<Flotsam>> &flotsam)
 			// Have all ships exit hyperspace at the same distance so that
 			// your escorts always stay with you.
 			double distance = (HYPER_C * HYPER_C) * .5 * HYPER_A + HYPER_D;
+			distance += currentSystem->ArrivalDistance();
```
Comment (tehhowch) [inline source/Ship.cpp]: So one thing to mention is that this is usually not the correct usage of the **`System`** arrival distance, because the `target` position is generally that of a planetary body, and not the system center. This would, however, be (part of) the correct usage of a **`Planet`**'s "HyperspaceExclusionRadius". Only when the player jumps into a system that has no planets with spaceports and has also not explicitly selected the planet in that system as a landing target would `target` be that of the system center (aka `Point { 0, 0 }`).

I guess the question is, "how correct do we want to be?" (Or, "how much vector math would you like to do?")

File: source/System.cpp (hunk 3 above)
Comment (Zitchas) [reply source/System.cpp]: What is the impact of negative values? Or I suppose the corollary question would be: Where does `arrivalDistance = 0` leave the player? Is it in the system center? or is is it the normal arrival spot around roughly the habitable zone?

Some speculations that come to mind:
If it is the former, and 0 is the center of the system, would a negative value catapult the player past the center of the system and drop them on the far side? If so, I'd say allow this value, because it make a neat change for certain specific systems or missions (If anyone were doing, say, a mission revolving around investigation of weird hyperspace phenomenon or something.)

What I'm getting at is this: So long as negative values have a noticeable and relatively consistent/predictable impact on ship arrival location, I would say we should allow them.

File: source/Ship.cpp (hunk 4 above)
Comment (Zitchas) [reply source/Ship.cpp]: Given the choice, I would prefer the hypserspace exclusion radius to be from system center rather from the planet. How complicated is the vector math involved?

File: source/System.cpp (hunk 3 above)
Comment (petervdmeer) [reply source/System.cpp]: `arrivalDistance = 0` is the default value, which leaves the player arriving at exactly the same spot as where he/she arrives without this PR.

A negative `arivalDistance` value would indeed catapult the player beyond his/her intended destination. I considered blocking negative values, but I actually came to the same conclusion as you; there are probably some content/plugin creators that can do something nice with negative values.

File: source/Ship.cpp (hunk 4 above)
Comment (petervdmeer) [reply source/Ship.cpp]: A simple solution would be to always use the system center as reference point when `arrivalDistance` is not zero. That should be a one-liner of code.

I didn't check yet how complicated the complex solution would be.

File: source/System.cpp (hunk 3 above)
Comment (Zitchas) [reply source/System.cpp]: Actually, if this gets merged, I'm going to include this in Remnant missions. :) A very long ways down the story tree, but still, fun ideas. : )

File: source/Ship.cpp (hunk 4 above)
Comment (petervdmeer) [reply source/Ship.cpp]: I implemented the simple solution with my last commit.

**Comment by petervdmeer** [comment]
Based on a good remark from @tehhowch on Discord I made a suggestion for multiple modifier keywords in #4334 .
The change here only affects hyperdrives, but if we accept the suggestion I made, then Zitchas can choose between 2 versions of the keywords in his plugin:
`arrival <distance>` for distances that should apply for both jump and hyperdrives in the future
and 
`arrival <distance> link` for distances that should only apply to hyperdrives in the future

**Comment by Zitchas** [comment]
Ah, so that's the difference between the two. Thanks for clarifying.

**Comment by geojak** [comment]
Help, sorry to bother you guys again. I would really love to use this feature in my plugins. Any chance of a merge for the next release?

**Comment by Amazinite** [comment]
> Any chance of a merge for the next release?

I'd like to have this merged for the next release (the associated issue is marked for v0.9.13). As for when it gets merged, I've been focusing on reviewing content PRs, so after that is when I'll be reviewing any mechanics/enhancements PRs. Tehhowch may get to this before I do, though.

File: source/System.h [hunk 5]
```diff
@@ -95,6 +95,8 @@ class System {
 	
 	// Get a list of systems you can travel to through hyperspace from here.
 	const std::set<const System *> &Links() const;
+	// Additional travel distance to target for ships entering through hyperspace.
+	double ArrivalDistance() const;
```
Comment (Pointedstick) [inline source/System.h:99]: To me, the function name does not match the description. I would expect that something called `ArrivalDistance` would be expressing an absolute distance, not an offset. A name like `AdditionalArrivalDistance` or `ArrivalOffset` might be clearer IMO.

Comment (petervdmeer) [reply source/System.h]: Changed to `ExtraArrivalDistance`.

**Comment by Amazinite** [comment]
Bumping so that I can link.

**Comment by petervdmeer** [comment]
Merged master to resolve minor merge conflict with jump distance PR.

**Review by Amazinite — COMMENTED** [review]
This doesn't seem to work with jump drives, which I'd expect to be the case with such a mechanic. (Having this work for jump drive would also be highly beneficial for #5560 to avoid jumping into the system right on top of the hazard.) From the RFC there is a mention of having keywords “link” or “jump” after the distance to specify only one or the other travel method being affected, so I’d expect a lack of either keyword to mean it affects both. 

Also, what would it take for the added distance to be relative to the target planet instead of forcing the distance to be from the system center? It seems odd to me that should this mechanic be used, some systems I would exit my jump near a planet and in other systems I would always seem to exit some set distance from the system center instead of near a planet.

Otherwise, seems to be working as intended right now with hyperdrive travel.

**Comment by petervdmeer** [comment]
This PR was just a small change to allow content creators to experiment with the more cinematic combat as mentioned in #4334.
I agree that we should also get this feature for jump drives and maybe I even create a follow-up PR for that. It just wasn't added into this PR.

The RFC mentions how to specify arrival distances from system-center and how to specify arrival distances from planet-centers (and also how to combine them). The basic distance comparisons are relatively easy (you just calculate all arrival distances from the center and from all planets that set distances and take the maximum one).
What could add some complexity is that a ship usually doesn't cross all objects that set arrival distance restrictions. If we want to go for a very precise calculation then we need to take into account the trajectory that a ship is on when entering the system relative to the objects that set arrival distance restrictions.

**Comment by Amazinite** [comment]
Well what I’m looking for right now is at the very least a mechanic to change the distance of arrival for both hyperdrive and jump drive travel. From where that distance is measured doesn’t really matter right now as far as I’m concerned. So how difficult would it be to have the arrival distance from this PR also apply to jump drives? Then in the future we could deal with customizing it further (e.g. hyperdrive only, jump drive only, distance relative to planets, distance relative to center, whatever).

**Comment by petervdmeer** [comment]
> <...> how difficult would it be to have the arrival distance from this PR also apply to jump drives?

The code change is very easy. But it is as easy to add here as it is to add in a separate PR.

I understand your need as content-creator, but this PR was created 1.5 years ago for a different need (more room for long range weapons in large space-battles) which I noticed Zitchas, geojak, ChamEV, Pointedstick, SirStigll and me are interested in.

If this PR is having bugs or issues, then I consider it very reasonable of you to ask me to fix those. But if this PR is working fine, and your only concern is that it is missing an additional feature that you really like to have, then I feel you should really review this PR for merging as-is (and hint during such review that you would really like to see another PR with the additional feature).
